## Contents
    01.Importing libraries
    02.Importing Data
    03.Data Wrangling Procedures
    04.Task 4.4 Steps

# 01.Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02.Importing Data

In [2]:
path=r'C:\Users\rache\Documents\Instacart Basket Analysis'

In [3]:
#importing data set order.csv
df_ords=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col=False)

In [4]:
df_ords = df_ords.drop(columns = ['eval_set'])

In [5]:
#importing data set products.csv
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

# 03.Data Wrangling Procedures

In [6]:
#checking for missing values
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [7]:
#renaming column order_dow
df_ords.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [8]:
#changing data type for columns that don't need descriptive statistics
df_ords['order_id'] = df_ords['order_id'].astype('str')
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [9]:
#checking data type change was successful
df_ords.dtypes

order_id                   object
user_id                    object
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [10]:
#importing data set departments.csv
df_dep=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col=False)

In [11]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [12]:
#redefining deparments dataframe in long form instead of wide
df_dept=df_dep.T

In [13]:
#fixing header row of 0
df_dept.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [14]:
#define the first row as the new header
new_header=df_dept.iloc[0]

In [15]:
#check new header
new_header

0    department
Name: department_id, dtype: object

In [16]:
#define new dataframe without first row
df_dept_new=df_dept[1:]

In [17]:
df_dept_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [18]:
#set new header as header for new dataframe
df_dept_new.columns = new_header

In [19]:
df_dept_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [20]:
#turning df_dept_new into data dictionary
data_dict = df_dept_new.to_dict('index')

In [21]:
#create new dataframe using data dictionary and products df
df_snacks=df_prods[df_prods['department_id']==19]

In [22]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

# 04.Task 4.4 Steps

## Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.
### Did this step earlier by turning user_id into string as well

## Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.

In [23]:
#Overwriting the column name for days_since_prior_order to days_since_last_order
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = True)

In [24]:
df_ords.columns

Index(['order_id', 'user_id', 'order_number', 'order_day_of_week',
       'order_hour_of_day', 'days_since_last_order'],
      dtype='object')

## Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [25]:
#pulling descriptive stats for order_hour_of_day column
df_ords['order_hour_of_day'].describe()

count    3.421083e+06
mean     1.345202e+01
std      4.226088e+00
min      0.000000e+00
25%      1.000000e+01
50%      1.300000e+01
75%      1.600000e+01
max      2.300000e+01
Name: order_hour_of_day, dtype: float64

In [26]:
#Changing the variable order_hour_of_day from a float to integer
df_ords['order_hour_of_day'] = df_ords['order_hour_of_day'].astype('int')

In [27]:
df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

### The busiest hour of the day is 10am with 288,418 orders

## Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [29]:
print(data_dict.get('4'))

{'department': 'produce'}


### Dept ID 4 is for produce

## The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [30]:
#creating a subset containing only breakfast item sales
df_breakfast=df_prods[df_prods['department_id']==14]

In [31]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

## They’d also like to see details about customers who might be throwing dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [33]:
#creating a subset containing only departments related to throwing dinner parties
df_dinner=df_prods[df_prods['department_id'].isin([5,20,7,12])]

## It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [34]:
df_dinner.shape

(7650, 5)

### df_dinner has 7,650 rows and 5 columns

## Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [38]:
#creating subset of only customer with user_id 1 from df_ords
df_ords_user_one=df_ords[df_ords['user_id']=='1']

In [39]:
df_ords_user_one

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   
5   3367565       1             6                  2                  7   
6    550135       1             7                  1                  9   
7   3108588       1             8                  1                 14   
8   2295261       1             9                  1                 16   
9   2550362       1            10                  4                  8   
10  1187899       1            11                  4                  8   

    days_since_last_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
5                    19.0  
6                    20.0  
7                    14.0  
8                     0.0  
9                    30.0  
10                   14.0

## You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [40]:
#details of user 1's behaviors
df_ords_user_one.describe()

order_number  order_day_of_week  order_hour_of_day  \
count     11.000000          11.000000          11.000000   
mean       6.000000           2.636364          10.090909   
std        3.316625           1.286291           3.477198   
min        1.000000           1.000000           7.000000   
25%        3.500000           1.500000           7.500000   
50%        6.000000           3.000000           8.000000   
75%        8.500000           4.000000          13.000000   
max       11.000000           4.000000          16.000000   

       days_since_last_order  
count              10.000000  
mean               19.000000  
std                 9.030811  
min                 0.000000  
25%                14.250000  
50%                19.500000  
75%                26.250000  
max                30.000000

In [42]:
#11 Exporting df_ords dataframe as “orders_wrangled.csv”
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [43]:
#12 Exporting df_dept_new dataframe as “departments_wrangled.csv”
df_dept_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))